### ISSUE TITLE
Macro Error on Serde Serialize/Deserialize: `Err` value: ParseFloatError { kind: Invalid } rust-analyzer macro-error

### ISSUE THREAD

**Issue by gbutler69** [rust-analyzer/rust-analyzer#8928]
With the latest Rust Analyzer release in VS Code (latest) I'm receiving the following error when deriving Serialize/Deserialize (Serde) on a simple struct:

> proc macro returned error: proc-macro panicked: called `Result::unwrap()` on an `Err` value: ParseFloatError { kind: Invalid } rust-analyzer [macro-error](https://rust-analyzer.github.io/manual.html#macro-error)

I am getting the error with the following simple code:

````rust
use std::{fs::File, io, path::Path};

use serde::{Deserialize, Serialize};

fn main() -> Result<(), io::Error> {
    let move_a = Move {
        direction: Direction::North,
        count: 3,
    };

    let file = serialize(&move_a, &Path::new("./tmp.txt"))?;
    let move_b = deserialize(file)?;

    println!("Move A = {:?}", move_a);
    println!("Move B = {:?}", move_b);

    assert_eq!(move_a, move_b);

    Ok(())
}

fn serialize(the_move: &Move, path: &Path) -> Result<File, io::Error> {
    serde_json::to_writer(io::BufWriter::new(File::create(path)?), the_move)?;
    File::open(path)
}

fn deserialize(file: File) -> Result<Move, io::Error> {
    let u = serde_json::from_reader(io::BufReader::new(file))?;
    Ok(u)
}

#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
enum Direction {
    North,
    NorthEast,
    East,
    SouthEast,
    South,
    SouthWest,
    West,
    NorthWest,
}

#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
struct Move {
    pub direction: Direction,
    pub count: u8,
}
````

This code compile as runs fine from the command line using ````cargo run````.

Also, somewhat counter-intuitively, the "Move" structure is recognized at the usage site as correctly implementing Serialize/Deserialize.

Is this a bug in rust-analyzer (it seems like it) or am I doing something wrong.

Here is my Cargo.toml:

````toml
[package]
name = "ex-bb2-serde-json"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
````

**Comment by matklad** [comment]
Hm, I cant' reproduce this using the latest version of rust-analyzer:

![image](https://user-images.githubusercontent.com/1711539/119237894-015f6300-bb48-11eb-82b0-b992994f4afd.png)

Cound you check the version of rust-analyzer you are using? https://rust-analyzer.github.io/manual.html#troubleshooting

**Comment by gbutler69** [comment]
The RA version is: v0.2.599. The VSCode version is: 1.56.2.

**Comment by lnicola** [comment]
And which `rustc` version?

**Comment by gbutler69** [comment]
Latest nightly.

**Comment by lnicola** [comment]
Dupe of #8925, I suppose.

**Comment by gbutler69** [comment]
NOTE: The error doesn't prevent the proper derivation of Serialize/Deserialize. It just shows the error (in the output tab if you don't have it configured to display errors inline). The error appears on the structure & enum. If I remove the Serialize/Deserialize from both, the error goes away, but then (obviously), Serialize/Deserialize is not implemented.

**Comment by gbutler69** [comment]
This does appear to be a nightly related issue. The problem goes away when I switch to stable.

**Comment by lnicola** [comment]
Unfortunately, we currently don't support multiple proc macro ABIs, so we can't fix this until 1.54 comes out.

**Comment by cynecx** [comment]
@gbutler69 You can use this [patch](https://github.com/cynecx/rust-analyzer/commit/1bf57ed78be76a75d926e01f1838c0e602af01a7) which applies the upstream abi changes on top of rust-analyzer. Note, that this doesn't implement the functionality of https://github.com/rust-lang/rust/pull/84717. But serde_derive should work without that.

**Comment by quantumsheep** [comment]
Same issue :(

**Comment by gbutler69** [comment]
> Same issue :(

@quantumsheep  Switch to stable for you project (if possible) using: rustup default stable.
If you need to use nightly, then you'll need to apply the patch and use a patched version of rust-analyzer as mentioned by @cynecx

**Comment by branpk** [comment]
If you need to use nightly but not necessarily the latest nightly, you can also use `rustup default nightly-2021-05-20`.

**Comment by sploders101** [comment]
I just got this issue while using Rust stable on mac (rustc 1.48.0). Read through this issue and thought "I'm not using nightly, but I'll try it anyway". Rust updated to 1.54.0 and the issue is gone, so it's also worthwhile to make sure stable Rust is up to date as well. Hope this helps someone else!

**Comment by tizoc** [comment]
I am experiencing this issue with `nightly-2021-08-04-x86_64-unknown-linux-gnu` in VSCode. I fixed it by overriding the rust-analyzer binary that is downloaded by VSCode with the one from the rustup toolchain install which gives me `rust-analyzer ea105f9 2021-08-03 dev`.

**Comment by flodiebold** [comment]
This was also fixed in #9550, I think.

@tizoc both of those versions are quite old. I'd instead recommend updating to the current stable version of rust-analyzer.

**Comment by tizoc** [comment]
@flodiebold the version from VSCode before I overwrote it was `rust-analyzer 3dae94bf2 2021-09-07 nightly` (after I switched to the "nightly" channel, didn't check before that but since the channel was the "stable" one, I suppose the downloaded binary was the latest "stable" too). But neither worked, I had to overwrite the binary.

I tried uninstalling and reinstalling the VSCode (from "stable" channel) first, then again but switching to the "nightly" channel.
